source/server/GameServer.py

```diff
@@ -104,14 +143,61 @@ def Send_scores(self):
 
     def Send_publicInfo(self):
         """Send the update to the melded cards on the table"""
-        #NOTE: visible_cards needs to be serialized.
-        #Current plan: never deserialize them, the client sends them in serialized and
-        #we leave them serialized in the channel during storage and thus when they go out again
-        self.Send_broadcast({"action": "publicInfo", "player_names": [p.name for p in self.players], "visible_cards": [p.visible_cards for p in self.players], "hand_status": [p.hand_status for p in self.players]})
+
+        #NOTE: visible_cards needs to be serialized form to be transmitted.
+        # On server keep them in serialized form.
+
+        if self.rules.Shared_Board:
+            # Shared_Board is True: (e.g. Liverpool) - each player can play on any players cards.
+            self.v_cards = [p.visible_cards for p in self.players]
+            if len(self.v_cards) == 0:
+                self.v_cards = [{}]
+            # v_cards contains a dictionary from each player/client
+            # each dictionary contains all the played cards that player/client is aware of.
+            # set self.visible_cards_now to the dictionary in v_cards with the most cards.
+            #
+            # todo: consider whether checking the length is the best way to determine which dictionary is most
+            # recent version of visible_cards.
+            max_len = -1
+            self.visible_cards_now = {}
```

Comment:
Why are you doing this calculation on broadcast instead of on receipt - when its updated players submit it to you and you can check if its longer than the one you currently have in self.visible_cards_now and update if it is, then when broadcasting you just broadcast self.visible_cards_now which you know is the largest you have recieved.

Reply:
Re-read code and realized I'm doing it this way because GamerServer is not bothering to check which player most recently updated board.  This does not optimize computation, but was simpler to code.

Reply:
Put this on todo list.

Reply:
Change made in branch: "Reduced frequency of vcards checks"